### ISSUE TITLE
way to control how pytest-xdist runs tests in parallel?

### ISSUE THREAD

**Issue by pytestbot** [pytest-dev/pytest#175]
Originally reported by: **Anonymous**

----------------------------------------

Apologies if this has been raised and triaged already. I could not find it in the archives. 

The enhancement I request in py.test is the ability to control how tests are executed in parallel. It would be _really_ nice to have a way to run multiple classes/modules in parallel, while having serial execution within a given set of classes/modules.

TestNG (Java) provides similar functionality. Please see:
http://testng.org/doc/documentation-main.html#parallel-running

This was also discussed a while ago on StackOverflow. Please see discussion:
http://stackoverflow.com/questions/4637036/is-there-a-way-to-control-how-pytest-xdist-runs-tests-in-parallel

----------------------------------------
- Bitbucket: https://bitbucket.org/pytest-dev/pytest/issue/175

**Comment by vladlaktionov** [comment]
it is nice to isolate all of the tests of course, but it is not always possible. Sometimes you really need some suites and i think it would be great to have such option in the pocket. Can't properly use xdist without it...

**Comment by pupssman** [comment]
So, is anything up here?

**Comment by nicoddemus** [comment]
Unfortunately not right now.

I have some ideas and plan to write a proposal to see what people think. I will link it here once it is ready.

**Comment by tobixx** [comment]
Just want to raise my hand here also. My use case is a multiple stages system test where I use parametrized fixtures with class scope. In this constellation I'm unable to use xdist at all (becasue next test needs results from the prior one) but really need it due too the runtime of this tests. 

Btw. the [incremental testing example ](http://pytest.org/latest/example/simple.html#incremental-testing-test-steps) will of course also not work as expected while using xdist.

**Comment by RonnyPfannschmidt** [comment]
this needs a different internal Design in xdists, its down in my pipeline but not in a soon place

**Comment by nchammas** [comment]
I have a similar use case.

I have a parameterized, module-scoped fixture. I want any given permutation of the fixture to be spun up only once since it is expensive to create in both time and money.

However, I want as many different permutations of the fixture to be spun up in parallel as possible, since that saves the most time and enables the most number of tests to execute simultaneously.

I would welcome any changes to xdist that enable this use case or that offer a viable alternative.

**Comment by nchammas** [comment]
Not sure if I've read the [xdist source](https://bitbucket.org/pytest-dev/pytest-xdist/src/8e1f81f8113ad87c9fab6ac346e8e71ecd2dc1ad/xdist/dsession.py?at=default&fileviewer=file-view-default#dsession.py-143:144) correctly, but an interim solution that may be easier to offer is some kind of hook that we can use to define how we want to chunk tests up.

[Source](https://bitbucket.org/pytest-dev/pytest-xdist/src/8e1f81f8113ad87c9fab6ac346e8e71ecd2dc1ad/xdist/dsession.py?at=default&fileviewer=file-view-default#dsession.py-143:144):
> class LoadScheduling:
> ...
>     Then the collection gets devided (sic) up in chunks and chunks get submitted to nodes.

Maybe we can get a hook similar to [`pytest_collection_modifyitems`](https://pytest.org/latest/writing_plugins.html#_pytest.hookspec.pytest_collection_modifyitems) which lets us define a custom chunking of tests to be sent off to test runners.

Does that make sense? Would anyone else find such a hook useful? I'm still getting familiar with pytest's architecture and concepts, but I'm eager to figure out a short-term solution that doesn't get in the way of future changes to xdist.

**Comment by RonnyPfannschmidt** [comment]
Any hook we prematurely introduce has the potential to severely cripple doing a proper solution

That and my current time constraints mean its unlikely to happen unless someone tries tries and can handle needing to maintain a temporary fork in case we cannot acceptance his/her solution for strategic reasons

**Comment by nchammas** [comment]
No worries, I completely understand. I'll take a look at @heipark's solution [here](https://github.com/heipark/pytest-xdist/compare/974bd566c599dc6a9ea291838c6f226197208b46...master) and try to come up with something that works for me. If it seems generally applicable, I'll share with y'all so you can consider it for inclusion.

The approach I'm looking to take is to offer a hook that people can use to custom define how to chunk their tests up for distribution to test runners. That's it. Options like chunking by class or by module can be implemented on top of this functionality.

I'll get to work on this and share it when I have something that works.

**Comment by RonnyPfannschmidt** [comment]
there is a general plan to do this for collection time already (so different nodes collect different files)

the main problem is development time and design time (which is hard to get by for personal reasons)

**Comment by nicoddemus** [comment]
I put up pytest-dev/pytest-xdist#17 with some ideas, would love to get some feedback! :smile:

**Comment by nicoddemus** [comment]
Oops, missed @nchammas post! :sweat_smile: 

A hook was an idea discussed before, but it was difficult to design a hook that would cover all possible use cases we had at the time. Also, as @RonnyPfannschmidt mentioned, currently the `master` node only has the items' node ids to work with in deciding how to schedule the tests, so it is a limiting factor too (of course we could work on that front and change `xdist` so slaves provide more more information to the master node during collection). But I think it is a good idea and worth thinking pursuing.

The proposal I wrote is meant to allow more fine grained control on how the tests are run as I in my use case there were some classes that should have their tests executed in a single slave while other modules also had the same requirement, all in the same test suite.

Anyway, @nchammas feel free to drop a line if you want to discuss this further (here or at pytest-dev/pytest-xdist#17).

**Comment by nchammas** [comment]
No worries @nicoddemus. :) I have changed my position to [the one described here](https://github.com/pytest-dev/pytest/issues/175#issuecomment-112200644). I've commented on your proposal, and I guess we can continue the discussion there.

**Comment by RonnyPfannschmidt** [comment]
This issue is superseded by the already referenced ones in the xdist repo
